### ISSUE TITLE
A short syntax for v-bind .prop modifier

### ISSUE THREAD

**Issue by trusktr** [vuejs/vue#7582]
### What problem does this feature solve?
Make things shorter and easier to write.

### What does the proposed API look like?
`:foo="bar"` is a syntax shortcut for `v-bind:foo="bar"`.

Similarly, `.foo="bar"` would be a nice shortcut for `:foo.prop="bar"` (and the leading `.` even makes sense!).

**Comment by trusktr** [comment]
This would be useful with my Custom Elements, which accept either attributes or the same-named props, and adding a whole extra `.prop` is not the same amount of convenience as writing `:` for `v-bind`.

**Comment by trusktr** [comment]
Worth noting, an attribute prefixed with `.` is not legal in DOM, as `document.body.setAttribute('.foo', '')` will throw. So taking up this symbol in Vue templates doesn't get in the way of DOM usage.

**Comment by trusktr** [comment]
Although interestingly, if a DOM element starts with an attribute like `.foo="bar"` in the HTML string, then that attributes gets parsed in and is legal; it is only forbidden when set via `setAttribute`. f.e. this works:

```js
document.body.innerHTML = `<div .foo="bar">inspect me to see .foo attribute</div>`
```

**Comment by yyx990803** [comment]
That's a great idea. Will consider this in 2.6.

**Comment by trusktr** [comment]
What do you think of `.foo.bar.baz="..."` to set a path in the object?

In my case, I can do stuff like following with my elements:

```js
el.rotation = [30, 30, 30]

// or

el.rotation.y = 30
el.rotation.x = 30
el.rotation.z = 30
```

**Comment by yyx990803** [comment]
Anything more than a shorthand is beyond the original scope of this proposal. And no, I don't think magic path extensions are a good idea. That's like reaching into the child component's private state.

**Comment by trusktr** [comment]
It's interesting, because then, do you suggest that I refactor my custom elements to have attributes and properties like the following?

```html
<i-mesh .rotation="" .rotation-y="" .rotation-x="" .rotation-y="">
```

and

```js
el.rotation = [...];
el.rotationY = ...;
el.rotationX = ...;
el.rotationZ = ...;
```

It makes the JavaScript API less clean.

**Comment by yyx990803** [comment]
Or `<i-mesh .rotation="{ x, y, z }">`?

**Comment by trusktr** [comment]
Yeah, it's easy to work around.

But that one is less performant:

```html
<template>
  <!-- creates a new object every frame -->
  <i-mesh .rotation="{ x, y, z }"></i-mesh>
</template>
<script>
  export default {
    // more variables
    data: () => ({ x: 0, y: 0, z: 0 }),
    mounted() {
      // animate this.y at 60fps
    },
  }
</script>
```

vs

```html
<template>
  <!-- passes only a number every frame -->
  <i-mesh .rotation.y="y"></i-mesh>
</template>
<script>
  export default {
    data: () => ({ y: 0 }),
    mounted() {
      // animate this.y at 60fps
    },
  }
</script>
```

I'll just make more attributes on the custom elements, and use `.rotation-y="y"`. Plus, this has the benefit of expanding the attribute API which may make it more flexible in other libraries.

**Comment by trusktr** [comment]
> Plus, this makes it clear which attributes map directly to properties, without reaching into places that are not represented as attributes. Maybe this is a good thing!

Well, maybe it's moot, because it's also possible that `.property` exists on a custom element instance without a `property=""` attribute being part of the element's HTML API. So, for example, SSR would just have to send the data to the client (not as an attribute, but as JS) then during hydration just send the data to the element instance property. In this case, having a notation like `.rotation.y=""` would work the same.

**Comment by yyx990803** [comment]
Closed via d2902ca8
